src/tools/wasm-fuzz-lattices.cpp

```diff
@@ -0,0 +1,389 @@
+      }
+    }
+  }
+
+  // Given two input - output lattice pairs of a transfer function, checks if
+  // the transfer function is monotonic. If this is violated, then we print out
+  // the CFG block input which caused the transfer function to exhibit
+  // non-monotonic behavior.
+  template<typename Lattice>
+  void checkMonotonicity(Lattice& lattice,
+                         std::string transferFunctionName,
+                         const analysis::BasicBlock* cfgBlock,
+                         typename Lattice::Element& first,
+                         typename Lattice::Element& second,
+                         typename Lattice::Element& firstResult,
+                         typename Lattice::Element& secondResult) {
+    analysis::LatticeComparison beforeCmp = lattice.compare(first, second);
+    analysis::LatticeComparison afterCmp =
+      lattice.compare(firstResult, secondResult);
+
+    // Cases in which monotonicity is preserved.
+    if (beforeCmp == analysis::LatticeComparison::NO_RELATION) {
+      // If there is no relation in the first place, we can't expect anything.
+      return;
+    } else if (beforeCmp == analysis::LatticeComparison::LESS &&
+               (afterCmp == analysis::LatticeComparison::LESS ||
+                afterCmp == analysis::LatticeComparison::EQUAL)) {
+      // x < y and f(x) <= f(y)
+      return;
+    } else if (beforeCmp == analysis::LatticeComparison::GREATER &&
+               (afterCmp == analysis::LatticeComparison::GREATER ||
+                afterCmp == analysis::LatticeComparison::EQUAL)) {
+      // x > y and f(x) >= f(y)
+      return;
+    } else if (beforeCmp == analysis::LatticeComparison::EQUAL &&
+               afterCmp == analysis::LatticeComparison::EQUAL) {
+      // x = y and f(x) = f(y)
+      return;
+    }
+
+    std::stringstream ss;
+
+    ss << "The lattice elements ";
+    first.print(ss);
+    ss << " -> ";
+    firstResult.print(ss);
+    ss << " and ";
+    second.print(ss);
+    ss << " -> ";
+    secondResult.print(ss);
+    ss << " show that " << transferFunctionName
+       << " is not monotone when given the input:\n";
+    cfgBlock->print(ss);
+    ss << "\n";
+
+    Fatal() << ss.str();
+  }
+
+  // Helper function for checking the properties of lattices and transfer
+  // functiosn on a CFG of a randomly generated function. It does this by
+  // radomly generating three lattice elements, and then using them as input
+  // states for a CFG block on which the transfer function is applied.
+  template<typename Lattice, typename TransferFunction>
+  void check(analysis::CFG& cfg,
+             Lattice& lattice,
+             TransferFunction& transferFunction,
+             std::string transferFuncName) {
+    for (auto cfgIter = cfg.begin(); cfgIter != cfg.end(); ++cfgIter) {
+      typename Lattice::Element x = lattice.getRandom(rand);
+      typename Lattice::Element y = lattice.getRandom(rand);
+      typename Lattice::Element z = lattice.getRandom(rand);
```

Comment:
It might make sense to move everything that doesn't depend on `cfgIter` out of the loop to reduce the amount of work done for a single seed and make debugging easier. Otherwise, if one of these checks fails, you need to know both the seed and how many times this loop has executed to step into the failing check. You could even go as far as to separate `check` out into `checkLattice` and `checkTransferFunction`.

Reply:
In order to make the user less dependent on loop execution
- I only generate one triple of lattice elements per function instead of one triple per block. However, as the current lattices are dependent on function-specific information for initialization, I can't do less.
- I used the seed to generate per-function "seeds". These I expose to the user in error messages, as well as the name of failing functions. The user can use these seeds and function names to re-run specific function with a function-specific seed.

Doing the above is a compromise between making the tests deterministic enough for easy replication while keeping open the possibility to introduce more random tests for every module generated.